staging/src/k8s.io/apimachinery/pkg/util/httpstream/wsstream/conn.go

```diff
@@ -272,6 +277,78 @@ func (conn *Conn) handle(ws *websocket.Conn) {
 	}
 }
 
+// handleV5 handles websocket data according to "V5" of the remotecommand subprotocol. The main
+// differences between V4 and V5 "handle" function is 1) a half-close signal on STDIN channel closes
+// the STDIN channel, and 2) handling a heartbeat "ping" frame results in a response "pong" frame.
+func (conn *Conn) handleV5(ws *websocket.Conn) {
+	defer conn.Close()
+	conn.initialize(ws)
+
+	for {
+		conn.resetTimeout()
+		var messageType byte
+		codec := websocket.Codec{
+			// Unmarshals binary data into "v" if it is a byte array, storing the "messageType".
+			Unmarshal: func(msg []byte, payloadType byte, v interface{}) (err error) {
+				messageType = payloadType
+				switch data := v.(type) {
+				case *[]byte:
+					*data = msg
+					return nil
+				}
+				return websocket.ErrNotSupported
+			},
+			// Always send a "pong" message.
+			Marshal: func(v interface{}) (data []byte, payloadType byte, err error) {
+				return nil, websocket.PongFrame, nil
+			},
+		}
+		var data []byte
+		if err := codec.Receive(ws, &data); err != nil {
+			if err != io.EOF {
+				klog.Errorf("Error on socket receive: %v", err)
+			}
+			break
+		}
+		if messageType == websocket.PingFrame {
+			// Send "pong" response for heartbeat
+			if err := codec.Send(ws, nil); err != nil {
+				klog.Errorf("Error sending pong response in websocket heartbeat: %v", err)
+			}
+			continue
+		}
+		if messageType != websocket.BinaryFrame {
+			klog.Errorf("Expected binary frame for websocket receive: %v", messageType)
+			break
+		}
+		if len(data) == 0 {
+			continue
+		}
+		channel := data[0]
+		data = data[1:]
+		if int(channel) >= len(conn.channels) {
+			klog.V(6).Infof("Frame is targeted for a reader %d that is not valid, possible protocol error", channel)
+			continue
+		}
+		// Half-close the stream when receiving a frame with the STDIN channel id, but
+		// with empty message data. This frame is interpreted as a Close
+		// signal and closes STDIN stream on this server-side. This protocol allows
+		// the command to complete and the remaining output to be sent to the client
+		// before the WebSockets Connection is closed.
+		//
+		// See https://github.com/kubernetes/kubernetes/issues/89899#issuecomment-1132502190.
+		if int(channel) == remotecommand.StreamStdIn && len(data) == 0 {
+			klog.V(4).Infof("Received half-close signal from client; close STDIN stream")
+			conn.channels[remotecommand.StreamStdIn].Close()
+			continue // MUST continue and call Receive again.
```

Comment:
ok, I think that this is not correct IIUIC, the SPDY implementation what does it so send a RST stream to signal it is closing 

https://github.com/kubernetes/kubernetes/blob/10a12165de5ebb5da1dd73163f790e90b181cd88/vendor/github.com/moby/spdystream/stream.go#L182-L207

The websocket protocol has a well defined close states https://www.rfc-editor.org/rfc/rfc6455#page-29. mayb the problem is this library does not implement it (most probably)

Reply:
it seems gorilla does it right https://github.com/gorilla/websocket/blob/76ecc29eff79f0cedf70c530605e486fc32131d1/conn.go#L85-L113